torch/autograd/profiler.py

```diff
@@ -366,14 +366,47 @@ class record_function(ContextDecorator):
     """
     def __init__(self, name):
         self.name = name
+        # Whether or not we should run record function's end callbacks when exiting.
+        self.run_callbacks_on_exit = True
 
     def __enter__(self):
         self.handle = torch.ops.profiler._record_function_enter(self.name)
+        return self
 
     def __exit__(self, *args):
-        torch.ops.profiler._record_function_exit(self.handle)
+        if self.run_callbacks_on_exit:
+            torch.ops.profiler._record_function_exit(self.handle)
         return False
 
+    def _call_end_callbacks_on_future(self, fut):
+        """
+        _call_end_callbacks_on_future is meant to be used for profiling async
+        calls that return a future. Calling this function will extend recording
+        beyond this scope, until the future is satisfied. It is useful for profiling
+        the end to end time of asynchronous calls. This function should only be called
+        once to attach the callback onto the future, and will throw if called multiple
+        times.
+
+        Arguments:
+            fut: (torch.distributed.rpc.Future or torch._C.Future): future for which to schedule
```

Comment:
`torch._C.Future` is an internal type? Would I be correct this API doc here is meant for PyTorch developers not users as the this function is an internal function as well?

Reply:
Yes, this function isn't meant to be called by users, hence the underscore. It looks like the jit Future is typed as `torch._C.Future` in Python and `torch.jit.Future` after https://github.com/pytorch/pytorch/pull/27637, so I guess, technically, the type here also depends on whether we are calling this from a Script function or not as well.

Reply:
A test with this JIT future would run as part of `test_profiler_with_script_async_rpc`, since rpc.async returns a JIT future if invoked with a torchscript function.

Reply:
We also test this in the rpc jit tests now.